File: consilium/oxford.py
"""Oxford debate mode: structured for/against with rebuttals and verdict."""

import random
import time

from .models import (
    JUDGE_MODEL,
    SessionResult,
    is_thinking_model,
    query_model,
    sanitize_speaker_content,
)
from .prompts import (
    DOMAIN_CONTEXTS,
    OXFORD_MOTION_TRANSFORM,
    OXFORD_CONSTRUCTIVE_SYSTEM,
    OXFORD_REBUTTAL_SYSTEM,
    OXFORD_CLOSING_SYSTEM,
    OXFORD_JUDGE_PRIOR,
    OXFORD_JUDGE_VERDICT,
)


def _assign_sides(
    debaters: list[tuple[str, str, tuple[str, str] | None]],
) -> tuple[tuple[str, str, tuple[str, str] | None], tuple[str, str, tuple[str, str] | None]]:
    """Random side assignment for Oxford debate."""
    sides = list(debaters)
    random.shuffle(sides)
    return sides[0], sides[1]


def run_oxford(
    question: str,
    debaters: list[tuple[str, str, tuple[str, str] | None]],
    api_key: str,
    google_api_key: str | None = None,
    verbose: bool = True,
    persona: str | None = None,
    domain: str | None = None,
    motion_override: str | None = None,
) -> SessionResult:
    """Run Oxford debate: structured for/against with rebuttals and verdict. Returns SessionResult."""
    start_time = time.time()
    cost_accumulator: list[float] = []
    judge_model = JUDGE_MODEL
    judge_name = "Judge (Claude)"

    domain_context = DOMAIN_CONTEXTS.get(domain, "") if domain else ""
    transcript_parts = []

    def _persona_suffix() -> str:
        if persona:
            return f"\n\nContext about the person asking: {persona}"
        return ""

    if verbose:
        print("=" * 60)
        print("OXFORD DEBATE")
        print("=" * 60)
        print()

    # === Phase 1: MOTION ===
    if motion_override:
        motion = motion_override
    else:
        motion_messages = [
            {"role": "system", "content": OXFORD_MOTION_TRANSFORM.format(question=question)},
            {"role": "user", "content": question},
        ]

        if verbose:
            print("## Motion\n")
            if is_thinking_model(judge_model):
                print("(thinking...)", flush=True)

        motion = query_model(
            api_key, judge_model, motion_messages,
            max_tokens=100, stream=verbose, cost_accumulator=cost_accumulator,
        )
        motion = motion.strip().strip('"')

        if verbose and is_thinking_model(judge_model):
            print(motion)

        if verbose:
            print()

    transcript_parts.append(f"## Motion\n\n{motion}")

    # Assign sides randomly
    prop, opp = _assign_sides(debaters)
    prop_name, prop_model, prop_fallback = prop
    opp_name, opp_model, opp_fallback = opp

    if verbose:
        print(f"Proposition (FOR): {prop_name}")
        print(f"Opposition (AGAINST): {opp_name}")
        print()

    transcript_parts.append(f"**Proposition:** {prop_name} | **Opposition:** {opp_name}")

    # === Phase 2: PRIOR ===
    if verbose:
        print("## Prior\n")
        print(f"### {judge_name}")
        if is_thinking_model(judge_model):
            print("(thinking...)", flush=True)

    prior_messages = [
        {"role": "system", "content": OXFORD_JUDGE_PRIOR.format(motion=motion) + _persona_suffix()},
        {"role": "user", "content": f"Motion: {motion}"},
    ]

    prior_response = query_model(
        api_key, judge_model, prior_messages,
        max_tokens=200, stream=verbose, cost_accumulator=cost_accumulator,
    )

    if verbose and is_thinking_model(judge_model):
        print(prior_response)

    if verbose:
        print()

    transcript_parts.append(f"## Prior\n\n### {judge_name}\n{prior_response}")

    # === Phase 3: CONSTRUCTIVE ===
    if verbose:
        print("## Constructive Speeches\n")

    transcript_parts.append("## Constructive Speeches")

    # Proposition constructive
    prop_system = OXFORD_CONSTRUCTIVE_SYSTEM.format(
        name=prop_name, side="FOR", motion=motion,
    ) + _persona_suffix()

    if verbose:
        print(f"### {prop_name} (Proposition)")
        if is_thinking_model(prop_model):
            print("(thinking...)", flush=True)

    prop_constructive = query_model(
        api_key, prop_model, [
            {"role": "system", "content": prop_system},
            {"role": "user", "content": f"Argue FOR the motion: {motion}"},
        ],
        max_tokens=800, stream=verbose, cost_accumulator=cost_accumulator,
    )

    if verbose and is_thinking_model(prop_model):
        print(prop_constructive)

    if verbose:
        print()

    transcript_parts.append(f"### {prop_name} (Proposition)\n{prop_constructive}")

    # Opposition constructive
    opp_system = OXFORD_CONSTRUCTIVE_SYSTEM.format(
        name=opp_name, side="AGAINST", motion=motion,
    ) + _persona_suffix()

    if verbose:
        print(f"### {opp_name} (Opposition)")
        if is_thinking_model(opp_model):
            print("(thinking...)", flush=True)

    opp_constructive = query_model(
        api_key, opp_model, [
            {"role": "system", "content": opp_system},
            {"role": "user", "content": f"Argue AGAINST the motion: {motion}"},
        ],
        max_tokens=800, stream=verbose, cost_accumulator=cost_accumulator,
    )

    if verbose and is_thinking_model(opp_model):
        print(opp_constructive)

    if verbose:
        print()

    transcript_parts.append(f"### {opp_name} (Opposition)\n{opp_constructive}")

    # === Phase 4: REBUTTAL ===
    if verbose:
        print("## Rebuttals\n")

    transcript_parts.append("## Rebuttals")

    # Proposition rebuts opposition
    prop_rebuttal_system = OXFORD_REBUTTAL_SYSTEM.format(
        name=prop_name, side="FOR", motion=motion,
        opponent_argument=sanitize_speaker_content(opp_constructive),
    ) + _persona_suffix()

    if verbose:
        print(f"### {prop_name} (Proposition rebuttal)")
        if is_thinking_model(prop_model):
            print("(thinking...)", flush=True)

    prop_rebuttal = query_model(
        api_key, prop_model, [
            {"role": "system", "content": prop_rebuttal_system},
            {"role": "user", "content": "Rebut the opposition's arguments."},
        ],
        max_tokens=600, stream=verbose, cost_accumulator=cost_accumulator,
    )

    if verbose and is_thinking_model(prop_model):
        print(prop_rebuttal)

    if verbose:
        print()

    transcript_parts.append(f"### {prop_name} (Proposition rebuttal)\n{prop_rebuttal}")

    # Opposition rebuts proposition
    opp_rebuttal_system = OXFORD_REBUTTAL_SYSTEM.format(
        name=opp_name, side="AGAINST", motion=motion,
        opponent_argument=sanitize_speaker_content(prop_constructive),
    ) + _persona_suffix()

    if verbose:
        print(f"### {opp_name} (Opposition rebuttal)")
        if is_thinking_model(opp_model):
            print("(thinking...)", flush=True)

    opp_rebuttal = query_model(
        api_key, opp_model, [
            {"role": "system", "content": opp_rebuttal_system},
            {"role": "user", "content": "Rebut the proposition's arguments."},
        ],
        max_tokens=600, stream=verbose, cost_accumulator=cost_accumulator,
    )

    if verbose and is_thinking_model(opp_model):
        print(opp_rebuttal)

    if verbose:
        print()

    transcript_parts.append(f"### {opp_name} (Opposition rebuttal)\n{opp_rebuttal}")

    # === Phase 5: CLOSING ===
    if verbose:
        print("## Closing Statements\n")

    transcript_parts.append("## Closing Statements")

    # Proposition closing
    prop_closing_system = OXFORD_CLOSING_SYSTEM.format(
        name=prop_name, side="FOR", motion=motion,
    )

    if verbose:
        print(f"### {prop_name} (Proposition closing)")
        if is_thinking_model(prop_model):
            print("(thinking...)", flush=True)

    prop_closing = query_model(
        api_key, prop_model, [
            {"role": "system", "content": prop_closing_system},
            {"role": "user", "content": "Give your closing statement."},
        ],
        max_tokens=400, stream=verbose, cost_accumulator=cost_accumulator,
    )

    if verbose and is_thinking_model(prop_model):
        print(prop_closing)

    if verbose:
        print()

    transcript_parts.append(f"### {prop_name} (Proposition closing)\n{prop_closing}")

    # Opposition closing
    opp_closing_system = OXFORD_CLOSING_SYSTEM.format(
        name=opp_name, side="AGAINST", motion=motion,
    )

    if verbose:
        print(f"### {opp_name} (Opposition closing)")
        if is_thinking_model(opp_model):
            print("(thinking...)", flush=True)

    opp_closing = query_model(
        api_key, opp_model, [
            {"role": "system", "content": opp_closing_system},
            {"role": "user", "content": "Give your closing statement."},
        ],
        max_tokens=400, stream=verbose, cost_accumulator=cost_accumulator,
    )

    if verbose and is_thinking_model(opp_model):
        print(opp_closing)

    if verbose:
        print()

    transcript_parts.append(f"### {opp_name} (Opposition closing)\n{opp_closing}")

    # === Phase 6: VERDICT ===
    if verbose:
        print("## Verdict\n")
        print(f"### {judge_name}")
        if is_thinking_model(judge_model):
            print("(thinking...)", flush=True)

    transcript_parts.append("## Verdict")

    debate_transcript = (
        f"### Proposition ({prop_name}) — Constructive\n{sanitize_speaker_content(prop_constructive)}\n\n"
        f"### Opposition ({opp_name}) — Constructive\n{sanitize_speaker_content(opp_constructive)}\n\n"
        f"### Proposition ({prop_name}) — Rebuttal\n{sanitize_speaker_content(prop_rebuttal)}\n\n"
        f"### Opposition ({opp_name}) — Rebuttal\n{sanitize_speaker_content(opp_rebuttal)}\n\n"
        f"### Proposition ({prop_name}) — Closing\n{sanitize_speaker_content(prop_closing)}\n\n"
        f"### Opposition ({opp_name}) — Closing\n{sanitize_speaker_content(opp_closing)}"
    )

    verdict_system = OXFORD_JUDGE_VERDICT.format(
        motion=motion,
        proposition_name=prop_name,
        opposition_name=opp_name,
        debate_transcript=debate_transcript,
    ) + _persona_suffix()

    verdict = query_model(
        api_key, judge_model, [
            {"role": "system", "content": verdict_system},
            {"role": "user", "content": f"Judge this debate on: {motion}\n\nYour prior assessment:\n{sanitize_speaker_content(prior_response)}"},
        ],
        max_tokens=1000, stream=verbose, cost_accumulator=cost_accumulator,
    )

    if verbose and is_thinking_model(judge_model):
        print(verdict)

    if verbose:
        print()

    transcript_parts.append(f"### {judge_name}\n{verdict}")

    duration = time.time() - start_time
    total_cost = round(sum(cost_accumulator), 4) if cost_accumulator else 0.0

    if verbose:
        print(f"({duration:.1f}s, ~${total_cost:.2f})")

    return SessionResult(
        transcript="\n\n".join(transcript_parts),
        cost=total_cost,
        duration=duration,
    )

File: src/modes/oxford.rs
//! Oxford debate mode: structured for/against with rebuttals and verdict.

use crate::api::{query_model, run_parallel_with_different_messages};
use crate::config::{
    resolved_judge_model, sanitize_speaker_content, CostTracker, Message, ModelEntry, SessionResult,
};
use crate::prompts::{
    oxford_closing_system, oxford_constructive_system, oxford_judge_prior, oxford_judge_verdict,
    oxford_motion_transform, oxford_rebuttal_system,
};
use crate::session::Output;
use rand::seq::SliceRandom;
use reqwest::Client;
use std::time::Instant;

pub async fn run_oxford(
    question: &str,
    models: &[ModelEntry],
    api_key: &str,
    google_api_key: Option<&str>,
    zhipu_api_key: Option<&str>,
    moonshot_api_key: Option<&str>,
    xai_api_key: Option<&str>,
    motion_override: Option<String>,
    _format: &str,
    timeout: f64,
    output: &mut dyn Output,
) -> SessionResult {
    let start = Instant::now();
    let cost_tracker = CostTracker::new();
    let client = Client::new();
    let judge_model = resolved_judge_model();

    let mut transcript_parts = Vec::new();

    let _ = output.write_str("============================================================\n");
    let _ = output.write_str("OXFORD DEBATE\n");
    let _ = output.write_str("============================================================\n\n");

    // Phase 1: MOTION
    let _ = output.begin_phase("MOTION");
    let motion = if let Some(m) = motion_override {
        m
    } else {
        let motion_t0 = Instant::now();
        let _ = output.begin_participant("Judge (Claude)");
        let _ = output.write_str("## Motion\n\n");
        let m_messages = vec![
            Message::system(oxford_motion_transform(question)),
            Message::user(question),
        ];
        let m = query_model(
            &client,
            api_key,
            judge_model.as_str(),
            &m_messages,
            100,
            timeout,
            2,
            Some(&cost_tracker),
        )
        .await;
        let m = m.trim().trim_matches('"').to_string();
        let _ = output.write_str(&format!("{}\n\n", m));
        let _ =
            output.end_participant("Judge (Claude)", &m, motion_t0.elapsed().as_millis() as u64);
        m
    };
    transcript_parts.push(format!("## Motion\n\n{motion}"));

    // Random side assignment
    let mut sides = models.to_vec();
    let mut rng = rand::thread_rng();
    sides.shuffle(&mut rng);

    let (prop_name, prop_model, prop_fallback) = sides[0];
    let (opp_name, opp_model, opp_fallback) = sides[1];

    let _ = output.write_str(&format!("Proposition (FOR): {prop_name}\n"));
    let _ = output.write_str(&format!("Opposition (AGAINST): {opp_name}\n\n"));
    transcript_parts.push(format!(
        "**Proposition:** {prop_name} | **Opposition:** {opp_name}"
    ));

    // Phase 2: PRIOR
    let _ = output.begin_phase("PRIOR");
    let prior_t0 = Instant::now();
    let _ = output.begin_participant("Judge (Claude)");
    let _ = output.write_str("## Prior\n### Judge (Claude)\n");
    let prior_messages = vec![
        Message::system(oxford_judge_prior(&motion)),
        Message::user(format!("Motion: {motion}")),
    ];
    let prior_response = query_model(
        &client,
        api_key,
        judge_model.as_str(),
        &prior_messages,
        200,
        timeout,
        2,
        Some(&cost_tracker),
    )
    .await;
    let _ = output.write_str(&format!("{}\n\n", prior_response));
    let _ = output.end_participant(
        "Judge (Claude)",
        &prior_response,
        prior_t0.elapsed().as_millis() as u64,
    );
    transcript_parts.push(format!("## Prior\n\n### Judge (Claude)\n{prior_response}"));

    // Phase 3: CONSTRUCTIVE (parallel)
    let _ = output.begin_phase("CONSTRUCTIVE SPEECHES");
    let _ = output.write_str("## Constructive Speeches\n(both sides arguing in parallel...)\n");
    let prop_system = oxford_constructive_system(prop_name, "FOR", &motion);
    let opp_system = oxford_constructive_system(opp_name, "AGAINST", &motion);

    let constructive_results = run_parallel_with_different_messages(
        &[
            (prop_name, prop_model, prop_fallback),
            (opp_name, opp_model, opp_fallback),
        ],
        &[
            vec![
                Message::system(prop_system),
                Message::user(format!("Argue FOR the motion: {motion}")),
            ],
            vec![
                Message::system(opp_system),
                Message::user(format!("Argue AGAINST the motion: {motion}")),
            ],
        ],
        api_key,
        google_api_key,
        zhipu_api_key,
        moonshot_api_key,
        xai_api_key,
        800,
        Some(&cost_tracker),
        None,
    )
    .await;

    let prop_constructive = constructive_results[0].2.clone();
    let opp_constructive = constructive_results[1].2.clone();
    for (name, _, response) in &constructive_results {
        let _ = output.write_str(&format!("### {name}\n{response}\n\n"));
        let _ = output.end_participant(name, response, 0);
    }

    transcript_parts.push(format!(
        "## Constructive Speeches\n\n### {prop_name} (Proposition)\n{prop_constructive}"
    ));
    transcript_parts.push(format!("### {opp_name} (Opposition)\n{opp_constructive}"));

    // Phase 4: REBUTTAL (parallel)
    let _ = output.begin_phase("REBUTTALS");
    let _ = output.write_str("## Rebuttals\n(both sides rebutting in parallel...)\n");
    let prop_rebuttal_system = oxford_rebuttal_system(
        prop_name,
        "FOR",
        &motion,
        &sanitize_speaker_content(&opp_constructive),
    );
    let opp_rebuttal_system = oxford_rebuttal_system(
        opp_name,
        "AGAINST",
        &motion,
        &sanitize_speaker_content(&prop_constructive),
    );

    let rebuttal_results = run_parallel_with_different_messages(
        &[
            (prop_name, prop_model, prop_fallback),
            (opp_name, opp_model, opp_fallback),
        ],
        &[
            vec![
                Message::system(prop_rebuttal_system),
                Message::user("Rebut the opposition's arguments."),
            ],
            vec![
                Message::system(opp_rebuttal_system),
                Message::user("Rebut the proposition's arguments."),
            ],
        ],
        api_key,
        google_api_key,
        zhipu_api_key,
        moonshot_api_key,
        xai_api_key,
        600,
        Some(&cost_tracker),
        None,
    )
    .await;

    let prop_rebuttal = rebuttal_results[0].2.clone();
    let opp_rebuttal = rebuttal_results[1].2.clone();
    for (name, _, response) in &rebuttal_results {
        let _ = output.write_str(&format!("### {name}\n{response}\n\n"));
        let _ = output.end_participant(name, response, 0);
    }

    transcript_parts.push(format!(
        "## Rebuttals\n\n### {prop_name} (Proposition rebuttal)\n{prop_rebuttal}"
    ));
    transcript_parts.push(format!(
        "### {opp_name} (Opposition rebuttal)\n{opp_rebuttal}"
    ));

    // Phase 5: CLOSING (parallel)
    let _ = output.begin_phase("CLOSING STATEMENTS");
    let _ = output.write_str("## Closing Statements\n(both sides closing in parallel...)\n");
    let prop_closing_system = oxford_closing_system(prop_name, "FOR", &motion);
    let opp_closing_system = oxford_closing_system(opp_name, "AGAINST", &motion);

    let closing_results = run_parallel_with_different_messages(
        &[
            (prop_name, prop_model, prop_fallback),
            (opp_name, opp_model, opp_fallback),
        ],
        &[
            vec![
                Message::system(prop_closing_system),
                Message::user("Give your closing statement."),
            ],
            vec![
                Message::system(opp_closing_system),
                Message::user("Give your closing statement."),
            ],
        ],
        api_key,
        google_api_key,
        zhipu_api_key,
        moonshot_api_key,
        xai_api_key,
        400,
        Some(&cost_tracker),
        None,
    )
    .await;

    let prop_closing = closing_results[0].2.clone();
    let opp_closing = closing_results[1].2.clone();
    for (name, _, response) in &closing_results {
        let _ = output.write_str(&format!("### {name}\n{response}\n\n"));
        let _ = output.end_participant(name, response, 0);
    }

    transcript_parts.push(format!(
        "## Closing Statements\n\n### {prop_name} (Proposition closing)\n{prop_closing}"
    ));
    transcript_parts.push(format!(
        "## Closing Statements\n\n### {opp_name} (Opposition closing)\n{opp_closing}"
    ));

    // Phase 6: VERDICT
    let verdict_t0 = Instant::now();
    let _ = output.begin_participant("Judge (Claude)");
    let _ = output.begin_phase("JUDGMENT");
    let _ = output.write_str("## Verdict\n### Judge (Claude)\n");
    let debate_transcript = format!(
        "### Proposition ({prop_name}) — Constructive\n{prop_constructive}\n\n### Opposition ({opp_name}) — Constructive\n{opp_constructive}\n\n### Proposition ({prop_name}) — Rebuttal\n{prop_rebuttal}\n\n### Opposition ({opp_name}) — Rebuttal\n{opp_rebuttal}\n\n### Proposition ({prop_name}) — Closing\n{prop_closing}\n\n### Opposition ({opp_name}) — Closing\n{opp_closing}",
        prop_constructive = sanitize_speaker_content(&prop_constructive),
        opp_constructive = sanitize_speaker_content(&opp_constructive),
        prop_rebuttal = sanitize_speaker_content(&prop_rebuttal),
        opp_rebuttal = sanitize_speaker_content(&opp_rebuttal),
        prop_closing = sanitize_speaker_content(&prop_closing),
        opp_closing = sanitize_speaker_content(&opp_closing)
    );

    let verdict_system = oxford_judge_verdict(&motion, prop_name, opp_name, &debate_transcript);
    let verdict_messages = vec![
        Message::system(verdict_system),
        Message::user(format!(
            "Judge this debate on: {motion}\n\nYour prior assessment:\n{prior_response}",
            prior_response = sanitize_speaker_content(&prior_response)
        )),
    ];

    let verdict = query_model(
        &client,
        api_key,
        judge_model.as_str(),
        &verdict_messages,
        1000,
        timeout,
        2,
        Some(&cost_tracker),
    )
    .await;

    let _ = output.write_str(&format!("{}\n\n", verdict));
    let _ = output.end_participant(
        "Judge (Claude)",
        &verdict,
        verdict_t0.elapsed().as_millis() as u64,
    );
    transcript_parts.push(format!("## Verdict\n\n### Judge (Claude)\n{verdict}"));

    let duration = start.elapsed().as_secs_f64();
    let cost = (cost_tracker.total() * 10000.0).round() / 10000.0;

    let _ = output.write_str(&format!("({:.1}s, ~${:.2})\n", duration, cost));

    SessionResult {
        transcript: transcript_parts.join("\n\n"),
        cost,
        duration,
        failures: None,
    }
}
